feat: add kube_node_status_addresses metric

stonith opened the pull request:
**What this PR does / why we need it**:
Adds `kube_node_status_addresses` metric to account for mutiple addresses per type.

**How does this change affect the cardinality of KSM**: *(increases, decreases or does not change cardinality)*
Adds a new metric per node

stonith commented on internal/store/node.go:
```diff
@@ -143,6 +143,15 @@ func createNodeInfoFamilyGenerator() generator.FamilyGenerator {
 
+			externalIP := ""
+			for _, address := range n.Status.Addresses {
+				if address.Type == "ExternalIP" {
```
I don't believe so, according to this test? https://github.com/kubernetes/kubernetes/blob/master/pkg/util/node/node_test.go#L50-L57

mrueg replied:
Thanks for checking!

I would suggest the following:

```

internalIP := ""
externalIP := ""
for _, address := range n.Status.Addresses {
    switch address.Type {
     case "InternalIP":
        internalIP = address.Address
     case "ExternalIP":
        externalIP = address.Address
    }
}

labelKeys = append(labelKeys, "internal_ip", "external_ip")
labelValues = append(labelValues, internalIP, externalIP)
```

mrueg replied:
Thanks again for testing. I'd assume in a dualstack env you would also be able to provide multiple external ips that the node is reachable under. This metric likely was added when k8s was still ipv4 only.

Probably best is that we'd freeze this metric and create a new `kube_node_address{type="{internal|external|hostname}", address="$address} 1` metric

CC: @dgrisonnet  @rexagod

stonith commented:
Seems to work as expected in a dualstack GKE cluster:

`kube_node_info{node="gke-dualstack-test-default-pool-b7e9ab8b-q7g4",kernel_version="5.15.109+",os_image="Container-Optimized OS from Google",container_runtime_version="containerd://1.7.0",kubelet_version="v1.27.3-gke.100",kubeproxy_version="v1.27.3-gke.100",provider_id="gce://test/us-west1-a/gke-dualstack-test-default-pool-b7e9ab8b-q7g4",pod_cidr="10.84.0.0/24",system_uuid="c880482c-39d8-8523-70de-0658b88a5fe6",internal_ip="10.0.0.3",internal_ip="2600:1900:4040:203:0:0:0:0",external_ip="35.203.180.214"} 1`

dgrisonnet commented:
I am not a fan of adding labels to the `_info` metric. Most of them are legacy and should be in their own metric. Ideally we should address that in the next major version since they are all marked as stable.

It would be better to have a dedicated  `kube_node_status_addresses` metric instead with type and ip labels. @stonith could you introduce this new metric instead of adding the new label?

@mrueg interested about your opinion on this

dgrisonnet commented:
Ah well I just saw https://github.com/kubernetes/kube-state-metrics/pull/2252#discussion_r1414582664, I guess our opinion align then :laughing:

mrueg commented:
Yep, the  kube_node_status_addresses would be my preference. Also I don't think prometheus allows to have the same label multiple times. 

Ideally it would look like this and we'll keep internal_ip still in there with a TODO to kill it in ksm-3.x
```
kube_node_status_address{type="internal", address="1.2.3.4"} 1
kube_node_status_address{type="internal", address="1::2"} 1
kube_node_status_address{type="external", address="2.3.4.5"} 1
kube_node_status_address{type="external", address="3::4"} 1
kube_node_status_address{type="hostname", address="example.com"} 1
```

stonith commented:
Updated PR to add kube_node_status_address metric

mrueg commented:
This looks good to me, two small requests:

* Can you add a comment to the _info metric e.g.`// TODO: remove internal_ip in v3, replaced by kube_node_status_address` ? 
* Can you update the metrics documentation?

stonith commented:
> This looks good to me, two small requests:
> 
> * Can you add a comment to the _info metric e.g.`// TODO: remove internal_ip in v3, replaced by kube_node_status_address` ?
> * Can you update the metrics documentation?

- Updated metric name to `kube_node_status_addresses` to be more accurate
- added new metric to the docs/node-metrics.md
- added comment to _info metric (I think that's where you meant?)

stonith commented:
added the missed `node` label

mrueg commented:
/lgtm

thanks!
